libfreerdp/locale/keyboard_layout.c

```diff
@@ -219,99 +219,96 @@ static const RDP_KEYBOARD_IME RDP_KEYBOARD_IME_TABLE[] =
 
 void freerdp_keyboard_layouts_free(RDP_KEYBOARD_LAYOUT* layouts)
 {
-    free(layouts);
+	RDP_KEYBOARD_LAYOUT* current = layouts;
+
+	if (!layouts)
+		return;
+
+	while ((current->code != 0) && (current->name != NULL))
+	{
+		free(current->name);
+		current ++;
+	}
+
+	free(layouts);
 }
 
 RDP_KEYBOARD_LAYOUT* freerdp_keyboard_get_layouts(DWORD types)
 {
-	int num, length, i;
+	size_t num, length, i;
 	RDP_KEYBOARD_LAYOUT* layouts;
 	RDP_KEYBOARD_LAYOUT* new;
-
 	num = 0;
 	layouts = (RDP_KEYBOARD_LAYOUT*) calloc((num + 1), sizeof(RDP_KEYBOARD_LAYOUT));
+
 	if (!layouts)
 		return NULL;
 
 	if ((types & RDP_KEYBOARD_LAYOUT_TYPE_STANDARD) != 0)
 	{
 		length = ARRAYSIZE(RDP_KEYBOARD_LAYOUT_TABLE);
 		new = (RDP_KEYBOARD_LAYOUT*) realloc(layouts, (num + length + 1) * sizeof(RDP_KEYBOARD_LAYOUT));
+
 		if (!new)
-		{
-			free(layouts);
-			return NULL;
-		}
+			goto fail;
 
 		layouts = new;
+
 		for (i = 0; i < length; i++, num++)
 		{
 			layouts[num].code = RDP_KEYBOARD_LAYOUT_TABLE[i].code;
 			layouts[num].name = _strdup(RDP_KEYBOARD_LAYOUT_TABLE[i].name);
-			if (!layouts[num].name)
-			{
-				for (--i; i >=0; --i)
-					free(layouts[num].name);
 
-				free(layouts);
-				return NULL;
-			}
+			if (!layouts[num].name)
+				goto fail;
 		}
 	}
+
 	if ((types & RDP_KEYBOARD_LAYOUT_TYPE_VARIANT) != 0)
 	{
 		length = ARRAYSIZE(RDP_KEYBOARD_LAYOUT_VARIANT_TABLE);
 		new = (RDP_KEYBOARD_LAYOUT*) realloc(layouts, (num + length + 1) * sizeof(RDP_KEYBOARD_LAYOUT));
+
 		if (!new)
-		{
-			free(layouts);
-			return NULL;
-		}
+			goto fail;
 
 		layouts = new;
+
 		for (i = 0; i < length; i++, num++)
 		{
 			layouts[num].code = RDP_KEYBOARD_LAYOUT_VARIANT_TABLE[i].code;
 			layouts[num].name = _strdup(RDP_KEYBOARD_LAYOUT_VARIANT_TABLE[i].name);
-			if (!layouts[num].name)
-			{
-				for (--i; i >=0; --i)
-					free(layouts[num].name);
 
-				free(layouts);
-				return NULL;
-			}
+			if (!layouts[num].name)
+				goto fail;
 		}
 	}
+
 	if ((types & RDP_KEYBOARD_LAYOUT_TYPE_IME) != 0)
 	{
 		length = ARRAYSIZE(RDP_KEYBOARD_IME_TABLE);
 		new = (RDP_KEYBOARD_LAYOUT*) realloc(layouts, (num + length + 1) * sizeof(RDP_KEYBOARD_LAYOUT));
+
 		if (!new)
-		{
-			free(layouts);
-			return NULL;
-		}
+			goto fail;
 
 		layouts = new;
+
 		for (i = 0; i < length; i++, num++)
 		{
 			layouts[num].code = RDP_KEYBOARD_IME_TABLE[i].code;
 			layouts[num].name = _strdup(RDP_KEYBOARD_IME_TABLE[i].name);
-			if (!layouts[num].name)
-			{
-				for (--i; i >=0; --i)
-					free(layouts[num].name);
 
-				free(layouts);
-				return NULL;
-			}
+			if (!layouts[num].name)
+				goto fail;
 		}
 	}
 
 	ZeroMemory(&layouts[num], sizeof(RDP_KEYBOARD_LAYOUT));
-
-	return layouts;
+	return &layouts[1];
```

Comment:
Why do we return ```&layouts[1]``` here? I think this will cause trouble when we call free() on this address ...

Reply:
@mfleisz Good you spottet that, missed to clean that one before commit.